### PR TITLE
planner: introduce a new variable `tidb_regard_null_as_point`

### PR DESCRIPTION
**Pull request by qw4990**
### First-time contributors' checklist 

- [x] I've signed [**Contributor License Agreement**](https://cla-assistant.io/pingcap/docs-cn) that's required for repo owners to accept my contribution.

### Which TiDB version(s) do your changes apply to? (Required)

**Tips for choosing the affected version(s):**

By default, **CHOOSE MASTER ONLY** so your changes will be applied to the next TiDB major or minor releases. If your PR involves a product feature behavior change or a compatibility change, **CHOOSE THE AFFECTED RELEASE BRANCH(ES) AND MASTER**.

For details, see [tips for choosing the affected versions (in Chinese)](https://github.com/pingcap/docs-cn/blob/master/CONTRIBUTING.md#版本选择指南).

- [x] master (the latest development version)
- [x] v5.4 (TiDB 5.4 versions)
- [ ] v5.3 (TiDB 5.3 versions)
- [ ] v5.2 (TiDB 5.2 versions)
- [ ] v5.1 (TiDB 5.1 versions)
- [ ] v5.0 (TiDB 5.0 versions)
- [ ] v4.0 (TiDB 4.0 versions)
- [ ] v3.1 (TiDB 3.1 versions)
- [ ] v3.0 (TiDB 3.0 versions)
- [ ] v2.1 (TiDB 2.1 versions)

### What is the related PR or file link(s)?

- This PR is translated from:
- Other reference link(s):

### Do your changes match any of the following descriptions?

- [ ] Delete files
- [ ] Change aliases
- [ ] Need modification after applied to another branch 
- [ ] Might cause conflicts after applied to another branch

### CODE REVIEW

File: system-variables.md [hunk 1]
```diff
@@ -678,6 +678,14 @@ MPP 是 TiFlash 引擎提供的分布式计算框架，允许节点之间的数
 - 默认情况下（即该变量值为 `ON` 时），某张表上的统计信息过期后，优化器认为该表上除总行数以外的统计信息不再可靠，转而使用 pseudo 统计信息。将该变量值设为 `OFF` 后，即使统计信息过期，优化器也仍会使用该表上的统计信息。
 - 如果表数据修改较频繁，没有及时对表执行 `ANALYZE`，但又希望执行计划保持稳定，可以将该变量值设为 `OFF`。
 
+### `tidb_regard_null_as_point` <span class="version-mark">从 v5.4.0 版本开始引入</span>
```
Comment (djshow832) [inline system-variables.md:681]: The system variables are sorted by name in this doc.

File: system-variables.md [hunk 2]
```diff
@@ -678,6 +678,14 @@ MPP 是 TiFlash 引擎提供的分布式计算框架，允许节点之间的数
...
+
+- 作用域：SESSION | GLOBAL
+- 默认值：`ON`
+- 这个变量用来控制优化器是否可以把 null 值当做点值并作为前缀条件来访问索引。
```
Comment (qiancai) [inline system-variables.md:685]: 请教下点值具体是什么值呢？搜了下 TiDB 文档，没出现过这个词

File: system-variables.md [hunk 3]
```diff
@@ -678,6 +678,14 @@ MPP 是 TiFlash 引擎提供的分布式计算框架，允许节点之间的数
...
+- 比如在有多列索引 `index(a, b)` 及查询条件 `a<=>null and b=1` 的情况下，如果关闭开关，则不能把 `b=1` 用来进行索引访问，因为 null 值不能被当做固定前缀。
```
Comment (qiancai) [inline system-variables.md:686]:
```suggestion
- 比如在有多列索引 `index(a, b)` 且查询条件中某列为空 `a<=>null and b=1` 的情况下，如果关闭该变量，则不能把 `b=1` 用来进行索引访问，因为 null 值不能被当做固定前缀。
```

File: system-variables.md (hunk 2 above)
Comment (xuyifangreeneyes) [reply system-variables.md:685]: 这里的点值跟优化器在索引上构造 scan range 有关。对于每个 range 来说，只有最后一列可以不为点值，而前面所有列都必须为点值。比如一段 range 为 `[a1 a2 ... an, b1 b2 ... bn]`，那么必然有 `a1 = b1`, ..., `a{n-1} = b{n-1}`，即前 n - 1 列都是点值。 在下面这个例子中，对于多列索引 `indx(a, b)`，如果 `tidb_regard_null_as_point` 为 `OFF`，构造的 range 是 `[null, null]`。如果 `tidb_regard_null_as_point` 为 `ON`，构造的 range 是 `[null 1, null 1]`。

File: system-variables.md (hunk 3 above)
Comment (xuyifangreeneyes) [reply system-variables.md:686]: 某列为空是指 `a<=>null` 吗

File: system-variables.md (hunk 2 above)
Comment (qw4990) [reply system-variables.md:685]: `这个变量用来控制优化器是否可以把包含 null 的等值条件作为前缀条件来访问索引` 这样呢 @qiancai  改成 “包含 null 的等值条件"

File: system-variables.md [hunk 4]
```diff
@@ -1301,6 +1301,13 @@ SET tidb_query_log_max_len = 20;
 - 这个变量用于控制在记录 TiDB 日志和慢日志时，是否将 SQL 中的用户信息遮蔽。
 - 将该变量设置为 `1` 即开启后，假设执行的 SQL 为 `insert into t values (1,2)`，在日志中记录的 SQL 会是 `insert into t values (?,?)`，即用户输入的信息被遮蔽。
 
+### `tidb_regard_null_as_point` <span class="version-mark">从 v5.4.0 版本开始引入</span>
+
+- 作用域：SESSION | GLOBAL
+- 默认值：`ON`
+- 这个变量用来控制优化器是否可以将包含 null 的等值条件作为前缀条件来访问索引。
+- 该变量默认开启，能减少需要访问的索引数据量，提高查询执行速度。例如，在有多列索引 `index(a, b)` 且查询条件为 `a<=>null and b=1` 的情况下，优化器可以同时使用查询条件的 `a<=>null` 和 `b=1` 进行索引访问。如果关闭该变量，因为 `a<=>null and b=1`包含 null 值的等值条件，优化器不会使用 `b=1` 进行索引访问。
```
Comment (qiancai) [inline system-variables.md:1309]:
```suggestion
- 该变量默认开启，能减少需要访问的索引数据量，提高查询执行速度。例如，在有多列索引 `index(a, b)` 且查询条件为 `a<=>null and b=1` 的情况下，优化器可以同时使用查询条件中的 `a<=>null` 和 `b=1` 进行索引访问。如果关闭该变量，因为 `a<=>null and b=1`包含 null 值的等值条件，优化器不会使用 `b=1` 进行索引访问。
```

File: system-variables.md [hunk 5]
```diff
@@ -1301,6 +1301,13 @@ SET tidb_query_log_max_len = 20;
 - 这个变量用于控制在记录 TiDB 日志和慢日志时，是否将 SQL 中的用户信息遮蔽。
 - 将该变量设置为 `1` 即开启后，假设执行的 SQL 为 `insert into t values (1,2)`，在日志中记录的 SQL 会是 `insert into t values (?,?)`，即用户输入的信息被遮蔽。
 
+### `tidb_regard_null_as_point` <span class="version-mark">从 v5.4.0 版本开始引入</span>
+
+- 作用域：SESSION | GLOBAL
+- 默认值：`ON`
+- 这个变量用来控制优化器是否可以将包含 null 的等值条件作为前缀条件来访问索引。
+- 该变量默认开启，能减少需要访问的索引数据量，提高查询执行速度。例如，在有多列索引 `index(a, b)` 且查询条件为 `a<=>null and b=1` 的情况下，优化器可以同时使用查询条件中的 `a<=>null` 和 `b=1` 进行索引访问。如果关闭该变量，因为 `a<=>null and b=1`包含 null 值的等值条件，优化器不会使用 `b=1` 进行索引访问。
```
Comment (qiancai) [inline system-variables.md:1309]:
```suggestion
- 该变量默认开启，能减少需要访问的索引数据量，提高查询执行速度。例如，在有多列索引 `index(a, b)` 且查询条件为 `a<=>null and b=1` 的情况下，优化器可以同时使用查询条件中的 `a<=>null` 和 `b=1` 进行索引访问。如果关闭该变量，因为 `a<=>null and b=1` 包含 null 值的等值条件，优化器不会使用 `b=1` 进行索引访问。
```

File: system-variables.md [hunk 6]
```diff
@@ -1301,6 +1301,13 @@ SET tidb_query_log_max_len = 20;
 - 这个变量用于控制在记录 TiDB 日志和慢日志时，是否将 SQL 中的用户信息遮蔽。
 - 将该变量设置为 `1` 即开启后，假设执行的 SQL 为 `insert into t values (1,2)`，在日志中记录的 SQL 会是 `insert into t values (?,?)`，即用户输入的信息被遮蔽。
 
+### `tidb_regard_null_as_point` <span class="version-mark">从 v5.4.0 版本开始引入</span>
+
+- 作用域：SESSION | GLOBAL
+- 默认值：`ON`
+- 这个变量用来控制优化器是否可以将包含 null 的等值条件作为前缀条件来访问索引。
+- 该变量默认开启，能减少需要访问的索引数据量，提高查询执行速度。例如，在有多列索引 `index(a, b)` 且查询条件为 `a<=>null and b=1` 的情况下，优化器可以同时使用查询条件中的 `a<=>null` 和 `b=1` 进行索引访问。如果关闭该变量，因为 `a<=>null and b=1` 包含 null 值的等值条件，优化器不会使用 `b=1` 进行索引访问。
```
Comment (qiancai) [inline system-variables.md:1309]:
```suggestion
- 该变量默认开启，能减少优化器需要访问的索引数据量，提高查询执行速度。例如，在有多列索引 `index(a, b)` 且查询条件为 `a<=>null and b=1` 的情况下，优化器可以同时使用查询条件中的 `a<=>null` 和 `b=1` 进行索引访问。如果关闭该变量，因为 `a<=>null and b=1` 包含 null 值的等值条件，优化器不会使用 `b=1` 进行索引访问。
```

**Review by djshow832 — APPROVED** [review]

**Review by xuyifangreeneyes — APPROVED** [review]

File: system-variables.md [hunk 7]
```diff
@@ -1301,6 +1301,13 @@ SET tidb_query_log_max_len = 20;
 - 这个变量用于控制在记录 TiDB 日志和慢日志时，是否将 SQL 中的用户信息遮蔽。
 - 将该变量设置为 `1` 即开启后，假设执行的 SQL 为 `insert into t values (1,2)`，在日志中记录的 SQL 会是 `insert into t values (?,?)`，即用户输入的信息被遮蔽。
 
+### `tidb_regard_null_as_point` <span class="version-mark">从 v5.4.0 版本开始引入</span>
+
+- 作用域：SESSION | GLOBAL
+- 默认值：`ON`
+- 这个变量用来控制优化器是否可以将包含 null 的等值条件作为前缀条件来访问索引。
+- 该变量默认开启，能减少优化器需要访问的索引数据量，提高查询执行速度。例如，在有多列索引 `index(a, b)` 且查询条件为 `a<=>null and b=1` 的情况下，优化器可以同时使用查询条件中的 `a<=>null` 和 `b=1` 进行索引访问。如果关闭该变量，因为 `a<=>null and b=1` 包含 null 值的等值条件，优化器不会使用 `b=1` 进行索引访问。
```
Comment (shichun-0415) [inline system-variables.md:1309]:
```suggestion
- 该变量默认开启。开启后，优化器需要访问的索引数据量减少，查询的执行速度提高。例如，在有多列索引 `index(a, b)` 且查询条件为 `a<=>null and b=1` 的情况下，优化器可以同时使用查询条件中的 `a<=>null` 和 `b=1` 进行索引访问。如果关闭该变量，因为 `a<=>null and b=1` 包含 null 的等值条件，优化器不会使用 `b=1` 进行索引访问。
```

**Review by shichun-0415 — APPROVED** [review]
rest LGTM

**Review by qiancai — APPROVED** [review]

File: system-variables.md [hunk 8]
```diff
@@ -1301,6 +1301,13 @@ SET tidb_query_log_max_len = 20;
 - 这个变量用于控制在记录 TiDB 日志和慢日志时，是否将 SQL 中的用户信息遮蔽。
 - 将该变量设置为 `1` 即开启后，假设执行的 SQL 为 `insert into t values (1,2)`，在日志中记录的 SQL 会是 `insert into t values (?,?)`，即用户输入的信息被遮蔽。
 
+### `tidb_regard_null_as_point` <span class="version-mark">从 v5.4.0 版本开始引入</span>
+
+- 作用域：SESSION | GLOBAL
+- 默认值：`ON`
+- 这个变量用来控制优化器是否可以将包含 null 的等值条件作为前缀条件来访问索引。
+- 该变量默认开启。开启后，该变量可以减少优化器需要访问的索引数据量，从而提高查询的执行速度。例如，在有多列索引 `index(a, b)` 且查询条件为 `a<=>null and b=1` 的情况下，优化器可以同时使用查询条件中的 `a<=>null` 和 `b=1` 进行索引访问。如果关闭该变量，因为 `a<=>null and b=1` 包含 null 的等值条件，优化器不会使用 `b=1` 进行索引访问。
```
Comment (qiancai) [inline system-variables.md:1309]:
```suggestion
- 该变量默认开启。开启后，该变量可以使优化器减少需要访问的索引数据量，从而提高查询的执行速度。例如，在有多列索引 `index(a, b)` 且查询条件为 `a<=>null and b=1` 的情况下，优化器可以同时使用查询条件中的 `a<=>null` 和 `b=1` 进行索引访问。如果关闭该变量，因为 `a<=>null and b=1` 包含 null 的等值条件，优化器不会使用 `b=1` 进行索引访问。
```

**Pull request merged**